### ISSUE TITLE
VRL: null and unicode escapes not working in string literals

### ISSUE THREAD

**Issue by hhromic** [vectordotdev/vector#11926]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

I recently found out that the null and unicode character escape sequences are not working in [string literal expressions in VRL](https://vector.dev/docs/reference/vrl/expressions/#string).
The documentation makes sense, therefore I think something is off in the VRL parser.

Examples from the VRL REPL:
```
$ "hello\0world"
error[E202]: syntax error
  ┌─ :1:1
  │
1 │ "hello\0world"
  │ ^^^^^^^^^^^^^^ unexpected error: invalid escape character: \0
  │
  = see language documentation at https://vrl.dev
```
```
$ "hello\u{1F30E}world"  # also tested with \U{1F30E} and \u1F30E just in case
error[E202]: syntax error
  ┌─ :1:1
  │
1 │ "hello\u{1F30E}world"
  │ ^^^^^^^^^^^^^^^^^^^^^ unexpected error: invalid escape character: \u
  │
  = see language documentation at https://vrl.dev
```

All the other documented escape sequences seems to be working fine (although not all printed correctly):
```
$ "hello\nworld"
"hello\nworld"

$ "hello\"world"
"hello\"world"

$ "hello\'world"
"hello'world"

$ "hello\\world"
"hello\\world"

$ "hello\nworld"
"hello\nworld"

$ "hello\rworld"  # \r (carriage-return) was rendered instead of printed with an escape
world"

$ "hello\tworld"  # \t (tab) was rendered instead of printed with an escape
"hello  world"
```

### Configuration

```text
N/A
```

### Version

vector 0.20.0 (x86_64-unknown-linux-gnu 2a706a3 2022-02-11)

**Comment by hhromic** [comment]
Looks like inded those escape characters are not implemented in the VRL lexer:

https://github.com/vectordotdev/vector/blob/7cb951daed57448b80463303a07184ed537d9c54/lib/vrl/parser/src/lex.rs#L1112-L1128

https://github.com/vectordotdev/vector/blob/7cb951daed57448b80463303a07184ed537d9c54/lib/vrl/parser/src/lex.rs#L1164-L1197

**Comment by hhromic** [comment]
After further investigation, it seems that these escape sequences used to work before, but they were not ported over in the VRL parser/compiler re-implementation PR #6353.

**Comment by jszwedko** [comment]
Interesting, thanks for the sleuthing! cc/ @StephenWakely @JeanMertz for thoughts on this.

**Comment by JeanMertz** [comment]
This is correct. I didn't port them over, not for any particular reason, other than wanting to start out simple, and adding more escape sequences as requested by the community.

I think it makes sense to handle escape characters similar to [how the Rust compiler handles them](https://doc.rust-lang.org/reference/tokens.html#ascii-escapes).

**Comment by hhromic** [comment]
I'm glad the omission was just for simplicity of the initial implementation instead of a limitation :)
And yes, having parity with Rust's escape character mechanism would be very nice and complete indeed.

In fairness, I don't think the missing escape characters I reported are a big use-case, I haven't seen anyone asking for them except for me in this issue after a long time since the new lexer was introduced.

The use-case for me that triggered discovering the missing escape characters is that I was testing my implementations of `ip_ntop()` and `ip_pton()` using the VRL REPL and I wasn't able to generate strings with arbitrary bytes. In other words, the escape I needed was the binary `\xNN` escape. Then the documentation pointed me to `\uNNNNNN` (which is not exactly the same but close-enough), which didn't work either.

**Comment by EdN-terascope** [comment]
In VRL, I'm having troubles with the limits Rust has for escaped characters and regex. PCRE2, Python or Golang style is preferable to Rust or ECMAScript here.

Thx
